sycl/doc/extensions/Barrier/Barrier.asciidoc

```diff
@@ -0,0 +1,290 @@
+
+This document describes an extension that adds `barrier`, which acts similarly to https://en.cppreference.com/w/cpp/thread/barrier[`std::barrier`].
+
+NOTE: CUDA can hardware accelerate from compute capability 8.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Revision: 1
+
+== Contact
+Tadej Ciglarič, Codeplay (tadej 'dot' ciglaric 'at' codeplay 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4.
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_CUDA_ASYNC_BARRIER` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+This extension introduces asynchronous barrier for CUDA devices. This extends `group_barrier` by splitting it into into two calls - arrive and wait. The wait call blocks until the predetermined number of work items in the same work group call arrive. This is also very similar to https://en.cppreference.com/w/cpp/thread/barrier[`std::barrier`] introduced in c++20.
+
+Implementing this requires some space in local memory, where the state of the barrier is kept (this is true even for the CUDA compute capability 8 that support this functionality in hardware).
+
+This extension introduces a `barrier` class, encapsulating the state of an asynchronous barrier. It must be used only in local memory, and reserves space in local memory even on hardware with native support (e.g. devices with CUDA Compute Capability 8.0). Once a barrier is initialized, it can be reused for multiple cycles of arrivals and waits.
+
+This extension can only be used on devices that support Independent Thread Scheduling (i.e. devices with CUDA Compute Capability 7.0 or higher) when using CUDA backend. On other backend the interface of `barrier` is not defined or declared. Attempting to call any functions defined in this extension with CUDA backend and a device with lower compute capability will cause kernel launch to fail.
```

Comment:
> On other backend the interface of `barrier` is not defined or declared. Attempting to call any functions defined in this extension with CUDA backend and a device with lower compute capability will cause kernel launch to fail.

The part about "`barrier` is not defined or declared" does not  make sense to me.  We don't know at compile-time which backend the kernel will be submitted to.  Therefore, it seems like the `barrier` class always needs to be declared, so long as the implementation supports this extension.

I think it would be more correct to say something like:

> The `barrier` class is an optional kernel feature as defined by the core SYCL specification section 5.7 "Optional kernel features".  This class may only be used by kernels that are submitted to a device that has the CUDA backend.  Attempting to submit a kernel using this class to another device causes the kernel invocation command (e.g. `parallel_for`) to  throw a synchronous exception with the `errc::kernel_not_supported` error code.

That wording assumes this class can run on *any* CUDA device.  If instead it only works on certain CUDA devices, we should introduce a new aspect, which applications can use to query whether the CUDA device supports this class.  In that case, the wording should be more like:

> The `barrier` class is an optional kernel feature as defined by the core SYCL specification section 5.7 "Optional kernel features".  This class may only be used by kernels that are submitted to a device that has the `aspect::ext_oneapi_XXX` aspect.  Attempting to submit a kernel using this class to another device causes the kernel invocation command (e.g. `parallel_for`) to  throw a synchronous exception with the `errc::kernel_not_supported` error code.

Reply:
It only works on certain devices, so I added an aspect and changed the specification to what is suggested. However I am not sure how to implement throwing a synchronous exception on launching a kernel using unsupported feature. If use the same implementation for device not supporting it, building the kernel will fail. Same will happen if libclc does not implement the functions for unsupported backend. If I instead call `__builtin_trap()` the error will happen asynchronously when kernel executes that instructions.

Reply:
The behavior with the synchronous exception will depend on the general [design for "optional device features"](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).  Once that design is in place, you will just need to decorate the `class barrier` definition with the appropriate C++ attribute.  That will avoid the compilation error and will cause the runtime to throw the synchronous exception.

Reply:
OK, I guess that means this extension needs to wait until that design is implemented. Or can we merge it as it is and add that attribute later?

Reply:
You do not need to wait for the optional device features design to be implemented.  As you say, we can add the C++ attribute later once that design is in place.  In the mean time, your extension will have the following limitations:

* Applications that target heterogeneous devices will need to be careful using your extension if one of the devices does not support this CUDA feature.  They may need to split kernels using your extension into a separate translation unit or link their application with `-fsycl-device-code-split=per_kernel` in order to avoid false compilation errors.

* If an application mistakenly submits a kernel using your extension to a device that does not support it, they may not get a synchronous exception as stated in the specification.  (They may get an online compilation failure or other error instead.)

Note that these limitations are similar for any "optional device feature".

Reply:
Added explanation of these limitations.

Reply:
Thanks for adding the explanation.  It would probably be better to add this in a "[NOTE]", though, which will emphasize that it is non-normative.

Reply:
Changed to a note.

Reply:
Thanks, done.

Reply:
Thanks!